[CSDiagnostics] Add method to retrieve information about an argument application

hamishknight opened the pull request:
This PR adds `FailureDiagnostic::getFunctionArgApplyInfo`, which retrieves information about an argument application at a given failure's locator. This method is then used to simplify the implementation of `NoEscapeFuncToTypeConversionFailure` and `MissingOptionalUnwrapFailure` and resolve a compiler crasher.


Resolves [SR-10811](https://bugs.swift.org/browse/SR-10811).

hamishknight committed:
Add FailureDiagnostic::resolveInterfaceType

This allows a raw type to be resolved using generic parameters where
possible.

hamishknight committed:
Add FailureDiagnostic::getFunctionArgApplyInfo

This allows failure diagnostics to retrieve information about the
application of an argument to a parameter at their locator.

hamishknight committed:
[AST][Sema] Add ValueDecl::hasCurriedSelf

This is a minimal cherry-pick of #25170.

hamishknight committed:
Store full locator info on ForceOptional

For the purposes of fix deduplication, we want to use the simplified
locator. However for the purposes of diagnostics, preserve the full
locator, which gives us more information to work with.

hamishknight commented:
cc. @xedin

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -110,6 +110,67 @@ Optional<SelectedOverload> FailureDiagnostic::getChoiceFor(Expr *expr) const {
+      callee = decl;
+
+    fnType = resolveType(overload->openedType)->getAs<AnyFunctionType>();
+  } else {
```
I think we should always get an overload, if we didn't then wrong locator was used? Because function applied to arguments supposed to be resolved down to declaration...

hamishknight replied:
@xedin Not if it's an arbitrary function expression such as `{ y in ... }(x)`.

xedin replied:
Ah right, I keep on forgetting that we allow that... It's kind of bad though because there should be a way to distinguish whether declaration should be there or not, might be worse adding an assert in there or something...

hamishknight replied:
@xedin Did you mean "might be worth adding an assert"? If so, did you have a specific condition in mind? The best I can think of is making sure the apply's function expr isn't a `DeclRefExpr`, `MemberRefExpr`, `UnresolvedDotExpr`, `UnresolvedMemberExpr`, etc. in the `else` branch. Though that's pretty suboptimal as it'll be easy to miss new cases as they arise.

xedin replied:
Yeah, sorry I meant "worth" :) We can go with what you suggest or I think it might be sufficient to check that function expr is a `ClosureExpr` and see if we could catch some more cases when declaration couldn't be found.

hamishknight replied:
@xedin Unfortunately I believe this can also arise for various things other than `ClosureExpr`, e.g `({ y in ... }, "").0(x)`, `curriedFn()(x)`, `[{y in ... }][0](x)` to name a few.

xedin replied:
I think all of those supposed to produce an "overload choice" at least e.g. `.0(x)` and `[0](x)` curried cases would as well, so maybe instead of using decl we should store OverloadChoice in new structure.

hamishknight replied:
@xedin Ah yes, you're right about `.0(x)`, I forgot we have overload choices for tuple elements!

For the others though, we shouldn't get overload choices from their "callee locators". This is especially important for the curried cases, as that could lead us to looking at the wrong parameter list. Actually I've just noticed a bug in `getCalleeLocator` that means we incorrectly return a subscript overload choice for the following:

```swift
struct S {
  subscript<T>() -> (T, T) -> Void { { _, _ in } }
}

func foo(_ x: @escaping () -> Int, _ y: () -> Int) {
  let s = S()
  _ = s[]({0}, y)
}
```

which leads to a crash because we're trying to get the second parameter from the subscript decl. What should've happened is we shouldn't have resolved a callee, and we should have worked with the function type of the apply instead.

xedin replied:
Yeah, that makes sense. Looks like curried calls as well as literal closure calls are good candidates for such check.

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -158,6 +160,12 @@ class FailureDiagnostic {
+  /// within an argument conversion, returns information about the application
+  /// of the argument to its parameter. If this failure is not for an argument
+  /// conversion, returns \c None.
+  Optional<FunctionArgApplyInfo> getFunctionArgApplyInfo() const;
```
I think it would be great for extensibility if this method would take a `ConstraintLocator *` as a parameter.

hamishknight replied:
@xedin Hah, that's actually how I originally had it, but wasn't sure whether or not it was worth it :) I'll go ahead and add that.

Thanks for the review!

xedin reviewed (approved):
Looks great, thank you! I have left a couple of minor comments inline.

hamishknight committed:
[CS] Don't retrieve an overload choice for x[](y)

Previously we returned a subscript member locator for an apply of a
subscript expr, which is incorrect because the callee is the function
returned from the subscript rather than the subscript itself.

hamishknight committed:
Use getFunctionArgApplyInfo in missing-unwrap diagnostic

hamishknight committed:
Use getFunctionArgApplyInfo in escaping diagnostics

This simplifies the implementation and fixes a compiler crasher.

Resolves SR-10811.

hamishknight committed:
Assert when we expect to resolve a callee overload

This helps ensure we don't miss cases where `getChoiceFor` was meant to
return a callee overload but doesn't.

hamishknight commented:
@xedin Thanks for the feedback! I've gone ahead and added an assertion to make sure we resolve a callee overload for most applies, and rebased to resolve a merge conflict.

In addition, I realised that the function interface type might not be a function type, for example if the callee is a var it could be a generic parameter type. Therefore I've changed the logic to accommodate this. Finally, I've added `FailureDiagnostic::resolveInterfaceType` which lets us properly resolve an interface type in the case where we don't have a callee.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -110,6 +110,161 @@ Optional<SelectedOverload> FailureDiagnostic::getChoiceFor(Expr *expr) const {
+
+      // Otherwise resolve its fixed type, mapped out of context.
+      if (auto fixed = cs.getFixedType(tvt))
+        return resolveInterfaceType(fixed->mapTypeOutOfContext());
```
I think it might make more sense to do `auto type = resolveType(tvt); return resolveInterfaceType(type->mapOutOfContext());` here because type variable could be bound to array where element was another type variable or something...

hamishknight replied:
@xedin I might be mistaken, but couldn't that cause us to "miss" a type variable that opens a generic parameter? For example if we have `$T0` bound to `[$T1]`, where `$T1` opens a generic placeholder `T`, and could be bound to say `Int`. If we call `resolveType` on `$T0` we'd get back `[Int]` rather than `[T]`.

That being said, I'm can't seem to find an example that does this, so it might not be a concern.

xedin replied:
Sorry yes, I misread this one, you want it to recurse and find my type variables! One more suggestion - you can simplify checking for generic parameter into `if (auto *GP = tvt->getImpl().getGenericParameter()) return GP;`

hamishknight replied:
@xedin Ah nice! Thanks, I've gone ahead and made that change.

xedin reviewed (approved):
Great! I have a small comment but otherwise I think this is ready to land.

hamishknight committed:
Simplify implementation of resolveInterfaceType a little

xedin commented:
@swift-ci please test

hamishknight commented:
Looks like the same unrelated Linux failure that https://github.com/apple/swift/pull/25170 faced.

xedin commented:
@swift-ci please test Linux platform

hamishknight commented:
Thanks @xedin! Is this (and/or https://github.com/apple/swift/pull/25170) good to go?

xedin commented:
Yes, absolute, sorry!

xedin commented:
@swift-ci please smoke test and merge

xedin commented:
@swift-ci please smoke test and merge
